axelor-sale/src/main/java/com/axelor/apps/sale/service/saleorder/SaleOrderLineTreeCalculationsServiceImpl.java

```diff
@@ -0,0 +1,132 @@
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+package com.axelor.apps.sale.service.saleorder;
+
+import com.axelor.apps.base.db.repo.ProductRepository;
+import com.axelor.apps.sale.db.SaleOrderLine;
+import com.axelor.apps.sale.db.SaleOrderLineTree;
+import com.axelor.apps.sale.db.repo.SaleOrderLineRepository;
+import com.axelor.apps.sale.db.repo.SaleOrderLineTreeRepository;
+import com.axelor.common.ObjectUtils;
+import com.google.inject.Inject;
+import com.google.inject.persist.Transactional;
+import java.math.BigDecimal;
+import java.util.List;
+
+public class SaleOrderLineTreeCalculationsServiceImpl
+    implements SaleOrderLineTreeCalculationsService {
+
+  protected SaleOrderLineTreeRepository saleOrderLineTreeRepository;
+  protected ProductRepository productRepository;
+  protected SaleOrderLineRepository saleOrderLineRepository;
+
+  @Inject
+  public SaleOrderLineTreeCalculationsServiceImpl(
+      SaleOrderLineTreeRepository saleOrderLineTreeRepository,
+      ProductRepository productRepository,
+      SaleOrderLineRepository saleorderLineRepository) {
+    this.saleOrderLineTreeRepository = saleOrderLineTreeRepository;
+    this.productRepository = productRepository;
+    this.saleOrderLineRepository = saleorderLineRepository;
+  }
+
+  @Override
+  @Transactional
+  public SaleOrderLine computePrices(SaleOrderLine saleOrderLine) {
+
+    List<SaleOrderLineTree> saleOrderLineTrees =
+        saleOrderLineTreeRepository
+            .all()
+            .filter("self.saleOrderLine= :saleOrderLineId")
+            .bind("saleOrderLineId", saleOrderLine.getId())
+            .fetch();
+
+    BigDecimal price =
+        saleOrderLineTrees.stream()
+            .map(this::computeSubPrices)
+            .reduce(BigDecimal.ZERO, BigDecimal::add);
+
+    saleOrderLine.setPrice(price);
+
+    saleOrderLineRepository.save(saleOrderLine);
+
+    return saleOrderLine;
+  }
+
+  protected BigDecimal computeSubPrices(SaleOrderLineTree tree) {
+
+    List<SaleOrderLineTree> saleOrderLineSubTrees =
+        saleOrderLineTreeRepository
+            .all()
+            .filter("self.parent = :parent")
+            .bind("parent", tree.getId())
+            .fetch();
+
+    if (ObjectUtils.isEmpty(saleOrderLineSubTrees)
+        && tree.getTypeSelect() == SaleOrderLineTreeRepository.PRODUCT_TYPE) {
+      return tree.getTotalPrice() != null ? tree.getTotalPrice() : BigDecimal.ZERO;
+    }
```

Comment:
```suggestion
    if (tree.getTypeSelect() == SaleOrderLineTreeRepository.PRODUCT_TYPE) {
      return Optional.ofNullable(tree.getTotalPrice()).orElse(BigDecimal.ZERO);
    }
```
If the type is `PRODUCT_TYPE` the sub-tree will always be empty. Otherwise, we can safely iterate on it because the method fetch above will never return a null list.

Reply:
ok